src/semaphore.c

```diff
@@ -716,6 +771,11 @@ _dispatch_thread_semaphore_create(void)
 	int ret = sem_init(s4, 0, 0);
 	DISPATCH_SEMAPHORE_VERIFY_RET(ret);
 	return (_dispatch_thread_semaphore_t)s4;
+#elif USE_FUTEX_SEM
+	dispatch_futex_t *futex = malloc(sizeof(*futex));
```

Comment:
it can have nested uses beacuse of dispatch_apply, that's why there is this `_semaphore_put()` thing to take it and leave it back in the TSD.

though you could use stack allocated stuff each time and I *think* it could work, and then nesting just works. it doesn't even really need to be in the TSD.

@das would have to confirm.